### PR TITLE
SF-1241: Fix issue where '&' character cannot be inserted into search box

### PR DESCRIPTION
**Pull request by jrmykolyn**

### CODE REVIEW

File: packages/@storefront/core/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] [patch]
+### Added
+- SF-1241: Expose table of DOM key strings.
+  - This table may be used by StoreFront components and/or implementations which must respond to DOM events.
```
Comment (JoeyDeol) [inline packages/@storefront/core/CHANGELOG.md]: Can you list the KEYS table here as well?

Comment (jrmykolyn) [reply packages/@storefront/core/CHANGELOG.md]: Noted, will do.

**Comment by jrmykolyn** [comment]
Adding a 'Do Not Merge' until this can be tested on IE*.

File: packages/@storefront/query/test/unit/search-box.ts [hunk 2]
```diff
@@ -29,25 +29,25 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
       });
 
       describe('onKeyDown()', () => {
-        it('should call event.preventDefault() if keyCode is up or down', () => {
+        it('should call event.preventDefault() if key is up or down', () => {
           const preventDefault = spy();
 
-          searchBox.state.onKeyDown(<any>{ keyCode: 38, preventDefault });
-          searchBox.state.onKeyDown(<any>{ keyCode: 40, preventDefault });
+          searchBox.state.onKeyDown(<any>{ key: 'ArrowUp', preventDefault });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 3]
```diff
@@ -29,25 +29,25 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
...
+          searchBox.state.onKeyDown(<any>{ key: 'ArrowDown', preventDefault });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 4]
```diff
@@ -29,25 +29,25 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
...
 
           expect(preventDefault).to.be.calledTwice;
         });
 
-        it('should not call event.preventDefault() if keyCode is any other key', () => {
+        it('should not call event.preventDefault() if key is any other key', () => {
           const preventDefault = () => expect.fail();
 
-          searchBox.state.onKeyDown(<any>{ keyCode: 13, preventDefault });
+          searchBox.state.onKeyDown(<any>{ key: 'Enter', preventDefault });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 5]
```diff
@@ -29,25 +29,25 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
...
         });
       });
 
       describe('onKeyUp()', () => {
         it('should set preventUpdate', () => {
-          const event: any = { keyCode: 10, target: {} };
+          const event: any = { key: 'ArrowUp', target: {} };
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 6]
```diff
@@ -63,7 +63,7 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
           searchBox.actions = <any>{ search };
           searchBox.services = <any>{ autocomplete: { hasActiveSuggestion: () => false } };
 
-          searchBox.state.onKeyUp(<any>{ keyCode: 13, target: { value } });
+          searchBox.state.onKeyUp(<any>{ key: 'Enter', target: { value } });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 7]
```diff
@@ -73,7 +73,7 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
           searchBox.flux = <any>{ emit };
           searchBox.services = <any>{ autocomplete: { hasActiveSuggestion: () => true } };
 
-          searchBox.state.onKeyUp(<any>{ keyCode: 13 });
+          searchBox.state.onKeyUp(<any>{ key: 'Enter' });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 8]
```diff
@@ -82,7 +82,7 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
           const emit = spy();
           searchBox.flux = <any>{ emit };
 
-          searchBox.state.onKeyUp(<any>{ keyCode: 27 });
+          searchBox.state.onKeyUp(<any>{ key: 'Escape' });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 9]
```diff
@@ -121,7 +121,7 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
           const emit = spy();
           searchBox.flux = <any>{ emit };
 
-          searchBox.state.onKeyUp(<any>{ keyCode: 40 });
+          searchBox.state.onKeyUp(<any>{ key: 'ArrowDown' });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/query/test/unit/search-box.ts [hunk 10]
```diff
@@ -130,7 +130,7 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
           const emit = spy();
           searchBox.flux = <any>{ emit };
 
-          searchBox.state.onKeyUp(<any>{ keyCode: 38 });
+          searchBox.state.onKeyUp(<any>{ key: 'ArrowUp' });
```
Comment (Vic-Dev) [inline packages/@storefront/query/test/unit/search-box.ts]: Import and use the `const`

File: packages/@storefront/structure/test/unit/filtered-list.ts [hunk 11]
```diff
@@ -87,7 +87,7 @@ suite('FilteredList', ({ expect, spy }) => {
   describe('onKeyDown()', () => {
     it('should select refinement', () => {
       const refinements: any = [{ value: 'a' , onClick: spy() }, { value: 'b', onClick: spy() }];
-      const keyboardEvent = { keyCode: 13 }
+      const keyboardEvent = { key: 'Enter' };
```
Comment (Vic-Dev) [inline packages/@storefront/structure/test/unit/filtered-list.ts]: Import and use the `const`

File: packages/@storefront/structure/test/unit/filtered-list.ts [hunk 12]
```diff
@@ -100,7 +100,7 @@ suite('FilteredList', ({ expect, spy }) => {
 
     it('should not select refinement', () => {
       const refinements: any = [{ value: 'a' , onClick: spy() }, { value: 'b', onClick: spy() }];
-      const keyboardEvent = { keyCode: 15 }
+      const keyboardEvent = { key: 'ArrowUp' }
```
Comment (Vic-Dev) [inline packages/@storefront/structure/test/unit/filtered-list.ts]: Import and use the `const`

File: packages/@storefront/core/src/core/utils.ts [hunk 13]
```diff
@@ -12,6 +12,13 @@ export const WINDOW = () => window;
 
 const ARRAY_TO_DOT_NOTATION_REGEX = /\[(\d+)\]/g;
 
+export const KEYS = {
```
Comment (jmbtutor) [inline packages/@storefront/core/src/core/utils.ts]: Does this table not exist somewhere already?

File: packages/@storefront/query/CHANGELOG.md [hunk 14]
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] [patch]
+### Changed
+- SF-1241: Update search box component to use table of DOM key strings.
+
+### Fixed
+- SF-1241: Fix Chrome-specific issue where '&' character cannot be inserted into search box.
```
Comment (jmbtutor) [inline packages/@storefront/query/CHANGELOG.md]: Backticks on `&`.

File: packages/@storefront/query/test/unit/search-box.ts (hunk 2 above)
Comment (jmbtutor) [inline packages/@storefront/query/test/unit/search-box.ts]: Should these use the exported constants?

Comment (jrmykolyn) [reply packages/@storefront/query/test/unit/search-box.ts]: I was thinking about this, and my concern is that using properties of the `KEYS` table will ensure that the functions under test are internally consistent (ie. that they account for those values in particular), but not that the values themselves are correct.

For example: if we changed the value of `KEYS.UP` from 'ArrowUp' to `__FOO__`, the test would still pass even though the value itself is invalid.

What do you think?

File: packages/@storefront/core/src/core/utils.ts (hunk 13 above)
Comment (jmbtutor) [reply packages/@storefront/core/src/core/utils.ts]: The latter, hoping for a built-in browser object or a library that we already have, but I guess one doesn't exist.

File: packages/@storefront/query/test/unit/search-box.ts (hunk 2 above)
Comment (jmbtutor) [reply packages/@storefront/query/test/unit/search-box.ts]: After some discussion with @jrmykolyn, I suggested that we still extract these values into constants, but keep the constants local to the file. These are standard browser values, so having them in two files should be fine.

File: packages/@storefront/query/test/unit/search-box.ts [hunk 15]
```diff
@@ -29,25 +39,29 @@ suite('SearchBox', ({ expect, spy, stub, itShouldConsumeAlias, itShouldProvideAl
...
-          searchBox.state.onKeyDown(<any>{ keyCode: 40, preventDefault });
+          searchBox.state.onKeyDown(<any>{ key: KEYS.UP, preventDefault });
+          searchBox.state.onKeyDown(<any>{ key: KEYS.DOWN, preventDefault });
+          // The following values are included for IE support.
+          // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
+          searchBox.state.onKeyDown(<any>{ key: KEYS.IE_UP, preventDefault });
+          searchBox.state.onKeyDown(<any>{ key: KEYS.IE_DOWN, preventDefault });
 
-          expect(preventDefault).to.be.calledTwice;
+          expect(preventDefault).to.be.callCount(4);
```
Comment (jmbtutor) [inline packages/@storefront/query/test/unit/search-box.ts]: `to.have.callCount(4)`

File: packages/@storefront/query/test/unit/search-box.ts (hunk 2 above)
Comment (Vic-Dev) [reply packages/@storefront/query/test/unit/search-box.ts]: just import the const pls

Comment (jrmykolyn) [reply packages/@storefront/query/test/unit/search-box.ts]: Updated test suite to import `KEYS` table from `core`. Resolving.

File: packages/@storefront/structure/test/unit/filtered-list.ts [hunk 16]
```diff
@@ -86,7 +87,7 @@ suite('FilteredList', ({ expect, spy, stub }) => {
 
   describe('onKeyDown()', () => {
     it('should exit early if the key is not Enter', () => {
-      const keyboardEvent: any = { keyCode: 1970 };
+      const keyboardEvent: any = { key: 1970 };
```
Comment (jmbtutor) [inline packages/@storefront/structure/test/unit/filtered-list.ts]: Make this a string. Perhaps use one of the existing ones.

Comment (jrmykolyn) [reply packages/@storefront/structure/test/unit/filtered-list.ts]: Implemented. Resolving.

**Pull request merged**